executor/runcexecutor/executor_linux.go

```diff
@@ -0,0 +1,308 @@
+	}
+
+	exitCh := reaper.Default.Subscribe()
+
+	eg, runCtx := errgroup.WithContext(ctx)
+	eg.Go(func() error {
+		_, err = w.runc.Run(runCtx, id, bundle, createOpts)
+		return err
+	})
+	eg.Go(func() error {
+		return <-initErrCh
+	})
+	err = eg.Wait()
+	if err != nil {
+		return containerd.UnknownExitStatus, errors.Wrap(err, "runc create failed")
+	}
+
+	pid, err := runc.ReadPidFile(pidfile.Name())
+	if err != nil {
+		return containerd.UnknownExitStatus, errors.Wrap(err, "failed to read runc pidfile")
+	}
+
+	status := waitpid(ctx, pid, exitCh)
+	defer endedOnce.Do(func() {
+		close(ended)
+	})
+	return status, nil
+}
+
+func (w *runcExecutor) exec(ctx context.Context, id, bundle string, pid0 int, specsProcess *specs.Process, process executor.ProcessInfo) (int, error) {
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	pidfile, err := ioutil.TempFile(bundle, "exec-*.pid")
+	if err != nil {
+		return containerd.UnknownExitStatus, errors.Wrap(err, "failed to create pidfile")
+	}
+	defer os.Remove(pidfile.Name())
+	pidfile.Close()
+
+	execOpts := &runc.ExecOpts{
+		PidFile: pidfile.Name(),
+		Detach:  true,
+	}
+	initErrCh := make(chan error)
+	if process.Meta.Tty {
+		socket, cleanup, err := w.setupConsole(ctx, initErrCh, nil, bundle, id, process)
+		if err != nil {
+			return containerd.UnknownExitStatus, errors.Wrap(err, "failed to setup console")
+		}
+		defer cleanup()
+		execOpts.ConsoleSocket = socket
+	} else {
+		close(initErrCh)
+		execOpts.IO = &forwardIO{stdin: process.Stdin, stdout: process.Stdout, stderr: process.Stderr}
+	}
+	exitCh := reaper.Default.Subscribe()
+	eg, execCtx := errgroup.WithContext(ctx)
+	eg.Go(func() error {
+		err = w.runc.Exec(execCtx, id, *specsProcess, execOpts)
+		return err
+	})
+	eg.Go(func() error {
+		return <-initErrCh
+	})
+	err = eg.Wait()
+	if err != nil {
+		return containerd.UnknownExitStatus, errors.Wrap(err, "runc create failed")
+	}
+
+	pid, err := runc.ReadPidFile(pidfile.Name())
+	if err != nil {
+		return containerd.UnknownExitStatus, errors.Wrap(err, "failed to read runc pidfile")
+	}
+
+	var wg sync.WaitGroup
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		// watch for parent process to die, if that happens then cancel the
+		// exec context
+		exitCh := reaper.Default.Subscribe()
+		waitpid(ctx, pid0, exitCh)
+		cancel()
+	}()
+	defer wg.Wait()
+
+	status := waitpid(ctx, pid, exitCh)
+	cancel() // stop the parent waitpid if still running
+	return status, nil
+}
+
+func (w *runcExecutor) setupConsole(ctx context.Context, initErrCh chan error, ended <-chan struct{}, dir, id string, process executor.ProcessInfo) (runc.ConsoleSocket, func(), error) {
+	cleaners := []func(){}
+	cleanOnReturn := true
+	defer func() {
+		if !cleanOnReturn {
+			return
+		}
+		for _, f := range cleaners {
+			f()
+		}
+	}()
+
+	socket, err := runc.NewTempConsoleSocket()
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "failed to create OCI runtime console socket")
+	}
```

Comment:
I think the code may be simpler and without the `initErrCh` if we just create the console socket and then handle everything below in one of the `(*errgroup.Group).Go` on line 103.

Then, the goroutines on line 235 and 253 can just be synchronous.

One more thing, the `cleaners` aren't returned on errors so: e.g. an error from `cio.NewFIFOSetInDir` means `socket.Close` won't get run.

Reply:
I think I see what you are saying, I will try to simplify.  Currently the cleaners are not returned on errors, they are run directly via a hacky defer, this will likely be simplified also with the refactoring you suggest.